clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1358,6 +1360,33 @@ mlir::Value CIRGenFunction::emitScalarConversion(mlir::Value src,
       .emitScalarConversion(src, srcTy, dstTy, loc);
 }
 
+mlir::Value ScalarExprEmitter::VisitUnaryLNot(const UnaryOperator *e) {
+  // Perform vector logical not on comparison with zero vector.
+  if (e->getType()->isVectorType() &&
+      e->getType()->castAs<VectorType>()->getVectorKind() ==
+          VectorKind::Generic) {
+    assert(!cir::MissingFeatures::vectorType());
+    cgf.cgm.errorNYI(e->getSourceRange(), "vector logical not");
+    return {};
+  }
+
+  // Compare operand to zero.
+  mlir::Value boolVal = cgf.evaluateExprAsBool(e->getSubExpr());
```

Comment:
This shouldn't be necessary AFAIK.  Clang should have already converted the expression via casts to bool.

Reply:
I wondered about that. This may be something we just blindly brought over from the classic codegen, which does the same thing. I'll try changing it there and run the regression tests to see if anything fails. If not, I'll remove it here.

Reply:
When I went to test the change, it became obvious why this is here, and I remembered that we had this same discussion a couple of weeks ago with the for-loop upstreaming. We have a boolean expression, but we need to emit a value. The AST will have an implicit cast to bool here, but `EvaluateExprAsBool()` does a couple of other things, like setting the current PGO statement, and doing RAII on FP operations.

I tried replacing this with emitScalarExpr in the classic codegen, and it does cause problems when the `!` operator is used with floating point values. I'm not clear why it doesn't handle the implicit cast in such cases, but it doesn't.